### PR TITLE
Compact data before streaming

### PR DESCRIPTION
**Pull request by denesb**
Currently, streaming and repair processes and sends data as-is. This is wasteful: streaming might be sending data which is expired or covered by tombstones, taking up valuable bandwidth and processing time. Repair additionally could be exposed to artificial differences, due to different nodes being in different states of compactness.
This PR adds opt-in compaction to `make_streaming_reader()`, then opts in all users. The main difference being in how these choose the current compaction time to use:
* Load'n'stream and streaming uses the current time on the local node.
* Repair uses a centrally chosen compaction time, generated on the repair master and propagated to al repair followers. This is to ensure all repair participants work with the exact state of compactness.

 Importantly, this compaction does *not* purge tombstones (tombstone GC is disabled completely).

Fixes: https://github.com/scylladb/scylladb/issues/3561

### CODE REVIEW

**Comment by denesb** [comment]
Don't merge yet, v2 is coming.

File: replica/database.hh [hunk 1]
```diff
@@ -686,21 +686,24 @@ public:
     //    reader and a _bounded_ amount of writes which arrive later.
     //  - Does not populate the cache
     // Requires ranges to be sorted and disjoint.
+    // When compaction_time is engaged, the reader's output will be compacted, with the provided query time.
+    // This compaction doesn't do tombstone garbage collection.
```
Comment (avikivity) [inline replica/database.hh:690]: Well, we should. Perhaps in a separate series.

File: replica/table.cc [hunk 2]
```diff
@@ -306,11 +319,14 @@ table::make_streaming_reader(schema_ptr s, reader_permit permit,
...
-    return make_flat_multi_range_reader(s, std::move(permit), std::move(source), ranges, slice, nullptr, mutation_reader::forwarding::no);
+    return maybe_compact_for_streaming(
+            make_flat_multi_range_reader(s, std::move(permit), std::move(source), ranges, slice, nullptr, mutation_reader::forwarding::no),
+            get_compaction_manager(),
+            compaction_time);
 }
 
 flat_mutation_reader_v2 table::make_streaming_reader(schema_ptr schema, reader_permit permit, const dht::partition_range& range,
-        const query::partition_slice& slice, mutation_reader::forwarding fwd_mr) const {
+        const query::partition_slice& slice, mutation_reader::forwarding fwd_mr, std::optional<gc_clock::time_point> compaction_time) const {
```
Comment (avikivity) [inline replica/table.cc:329]: We should make this overload call the other.

File: streaming/stream_transfer_task.cc [hunk 3]
```diff
@@ -70,7 +70,7 @@ struct send_info {
         , cf(tbl_)
         , ranges(std::move(ranges_))
         , prs(dht::to_partition_ranges(ranges))
-        , reader(cf.make_streaming_reader(cf.schema(), std::move(permit_), prs, {}))
+        , reader(cf.make_streaming_reader(cf.schema(), std::move(permit_), prs, gc_clock::now()))
         , update(std::move(update_fn))
```
Comment (avikivity) [inline streaming/stream_transfer_task.cc:74]: If the clocks on the other node have a significant lag, they'll observe different expiration times.

Concrete example:
 - this node has the clocks set in the future
 - other nodes have correct clocks
 - we stream to the other nodes
 - other nodes observe premature expiration

but, it's not like this patch is making anything worse; compaction could have kicked a little while ago and expired that data anyway. So no problem here.

File: message/messaging_service.cc [hunk 4]
```diff
@@ -1294,14 +1294,14 @@ future<> messaging_service::send_repair_put_row_diff(msg_addr id, uint32_t repai
 }
 
 // Wrapper for REPAIR_ROW_LEVEL_START
-void messaging_service::register_repair_row_level_start(std::function<future<repair_row_level_start_response> (const rpc::client_info& cinfo, uint32_t repair_meta_id, sstring keyspace_name, sstring cf_name, dht::token_range range, row_level_diff_detect_algorithm algo, uint64_t max_row_buf_size, uint64_t seed, unsigned remote_shard, unsigned remote_shard_count, unsigned remote_ignore_msb, sstring remote_partitioner_name, table_schema_version schema_version, rpc::optional<streaming::stream_reason> reason)>&& func) {
+void messaging_service::register_repair_row_level_start(std::function<future<repair_row_level_start_response> (const rpc::client_info& cinfo, uint32_t repair_meta_id, sstring keyspace_name, sstring cf_name, dht::token_range range, row_level_diff_detect_algorithm algo, uint64_t max_row_buf_size, uint64_t seed, unsigned remote_shard, unsigned remote_shard_count, unsigned remote_ignore_msb, sstring remote_partitioner_name, table_schema_version schema_version, rpc::optional<streaming::stream_reason> reason, rpc::optional<gc_clock::time_point> compaction_time)>&& func) {
```
Comment (avikivity) [inline message/messaging_service.cc:1297]: > Using a centrally generated compaction-time, generated on the repair
master and propagated to all repair followers. For repair it is
imperative that all participants use the exact same compaction time,
otherwise there can be artificial differences between participants,
generating unnecessary repair activity.
If the repair master is an old node, and hence doesn't send a
compaction-time to the followers, no compaction is done. If the repair
master is a new node, but some of the followers are an old node, this
will result in some participants compacting, while others not. This can
generate significant amount of extra repair activity. In my opinion we
can live with this, we don't recommend repairing mixed clusters anyway.

"some compacting and some not" isn't any different from the current behavior.

I'd like to decouple repair scheduling and cluster upgrades, but this patch doesn't harm it.

File: repair/row_level.cc [hunk 5]
```diff
@@ -2302,20 +2328,25 @@ future<> repair_service::init_ms_handlers() {
...
+                range, algo, max_row_buf_size, seed, remote_shard, remote_shard_count, remote_ignore_msb, schema_version, reason, compaction_time, this] (repair_service& local_repair) mutable {
             if (!local_repair._sys_dist_ks.local_is_initialized() || !local_repair._view_update_generator.local_is_initialized()) {
                 return make_exception_future<repair_row_level_start_response>(std::runtime_error(format("Node {} is not fully initialized for repair, try again later",
                         utils::fb_utilities::get_broadcast_address())));
             }
             streaming::stream_reason r = reason ? *reason : streaming::stream_reason::repair;
+            std::optional<gc_clock::time_point> ct;
+            if (compaction_time) {
+                ct = *compaction_time;
+            }
```
Comment (avikivity) [inline repair/row_level.cc:2345]: Why not use optional\<> throughout? "0 is a special case" is just more more thing to keep track of and possibly forget.

File: repair/row_level.hh [hunk 6]
```diff
@@ -214,6 +214,7 @@ public:
             shard_config master_node_shard_config,
             table_schema_version schema_version,
             streaming::stream_reason reason,
+            std::optional<gc_clock::time_point> compaction_time,
```
Comment (avikivity) [inline repair/row_level.hh:217]: Is there nothing in idl for this? we added verb support a long time ago.

Please sanity test manually mixed version repair in both directions.

**Comment by avikivity** [comment]
Looks good.

File: repair/row_level.cc (hunk 5 above)
Comment (denesb) [reply repair/row_level.cc:2345]: `compaction_time` is an `rpc::optional<>`. I cannot use `std::optional<>` in RPC, because RPC uses `rpc::optional<>` to denote parameter which might or might not be sent.

File: repair/row_level.hh (hunk 6 above)
Comment (denesb) [reply repair/row_level.hh:217]: Not sure what you mean. This is not the verb definition itself. But I also amended that one.

File: replica/database.hh (hunk 1 above)
Comment (denesb) [reply replica/database.hh:690]: This was a conscious decision. See the huge discussion on https://github.com/scylladb/scylladb/issues/3561 for more context.
TLDR: @vladzcloudius insists that we should do all we can do to prevent data resurrection, this includes sending expired tombstones. If we have it, we might as well send it as the remote could have data it covers.

File: streaming/stream_transfer_task.cc (hunk 3 above)
Comment (denesb) [reply streaming/stream_transfer_task.cc:74]: This was my thinking as well.

File: test/boost/mutation_reader_test.cc [hunk 7]
```diff
@@ -2310,7 +2310,7 @@ SEASTAR_THREAD_TEST_CASE(test_multishard_streaming_reader) {
                 return dht::to_partition_range(*next);
             }
             return std::nullopt;
-        });
+        }, {});
```
Comment (mykaul) [inline test/boost/mutation_reader_test.cc:2313]: Do we plan on adding more tests?

Comment (denesb) [reply test/boost/mutation_reader_test.cc:2313]: It would make sense. I just have to think of a way on how to do it.

File: repair/row_level.cc [hunk 8]
```diff
@@ -2302,20 +2328,25 @@ future<> repair_service::init_ms_handlers() {
...
             return repair_meta::repair_row_level_start_handler(local_repair, from, src_cpu_id, repair_meta_id, std::move(ks_name),
                     std::move(cf_name), std::move(range), algo, max_row_buf_size, seed,
                     shard_config{remote_shard, remote_shard_count, remote_ignore_msb},
-                    schema_version, r, _repair_module->abort_source());
+                    schema_version, r, ct, _repair_module->abort_source());
```
Comment (tgrabiec) [inline repair/row_level.cc:2349]: Didn't you forget to pass this to repair_reader::make_reader()?

Comment (denesb) [reply repair/row_level.cc:2349]: Yes, I did, hence https://github.com/scylladb/scylladb/pull/14756#issuecomment-1641767170 :).

File: repair/row_level.cc (hunk 5 above)
Comment (avikivity) [reply repair/row_level.cc:2345]: rpc::optional`<std::optional`<...>>?

File: replica/table.cc (hunk 2 above)
Comment (denesb) [reply replica/table.cc:329]: The returned reader out-lives the co-routine frame.

File: repair/row_level.hh (hunk 6 above)
Comment (avikivity) [reply repair/row_level.hh:217]: We have things like

```
verb [[with_client_info, cancellable]] direct_fd_ping (raft::server_id dst_id) -> service::direct_fd_ping_reply;
```

I thought we converted all verbs, but maybe not.

File: repair/row_level.cc (hunk 5 above)
Comment (denesb) [reply repair/row_level.cc:2345]: Actually, in v2 I move away from `std::optional<gc_clock::time_point>` completely and just use `gc_clock::time_point`.

File: repair/row_level.hh (hunk 6 above)
Comment (denesb) [reply repair/row_level.hh:217]: Looks like repair was not converted. @asias ?

**Comment by denesb** [comment]
v2:
* Actually propagate down compaction time to repair reader.
* Instead of opting-out of compaction when repair master is an old node, just generate a local compaction-time on followers if master didn't send one. With this, no opt-out mechanism is necessary anymore.
* Make compaction time mandatory argument in `make_[multishard_]streaming_reader()` to avoid mistakes like the above. All users want to opt-in anyway (see above).

File: repair/row_level.cc (hunk 8 above)
Comment (denesb) [reply repair/row_level.cc:2349]: Fixed in v2.

**Comment by denesb** [comment]
TODO: write tests (will have to be dtests).

File: replica/database.hh (hunk 1 above)
Comment (vladzcloudius) [reply replica/database.hh:691]: > That's why we have tombstone_gc = repair.

As mentioned in the referenced discussion, this doesn't solve all the problems and hence we should continue sending expired tombstones.
@avikivity note that this PR deals with the sending part of streaming (hence "before" in the name).

Expired tombstones can and should be dropped on the receiving side of the streaming. It think some work has already been done on the topic by @asias.

Bottom line: this part of the PR is good as-is and IMO no need to change that in the future.

**Review by raphaelsc — APPROVED** [review]
looks good

**Review by tgrabiec — APPROVED** [review]

File: repair/row_level.hh (hunk 6 above)
Comment (asias) [reply repair/row_level.hh]: Not all verbs are converted.

**Comment by asias** [comment]
Since expired tombstones are NOT dropping with the compacting reader, it is pretty safe. 

I have concerns that the compacting reader overhead introduced for the case there is nothing or little to compact and drop. This will make repair slower than before since we have to iterate all the cells in the stream to figure out nothing to compact and drop. 

For example, we  drop 1% of the data when compacting reader is turned on, however it might happen that the overhead is much bigger than not turning on the compacting reader.

In most cases, the normal compaction should have dropped the shadowed data. So the compacting reader is supposed to drop small part of the data. 

I think we'd better have a switch to turn on / off  the compacting reader. We could have some logic to turn it on and off automatically later, e.g., some estimating of the data shape (follow up). In this series, let's at least have a switch to disable compacting reader.

**Comment by denesb** [comment]
v3:
* Add config option allowing force opt-out from compaction.
* Wrote a dtest which verifies that compaction happens byte default and doesn't happen with the option turned off: https://github.com/scylladb/scylla-dtest/pull/3321
* Manual upgrade test: mixed cluster of 2 nodes (one master, one with this PR). Repair started on either nodes finishes successfully.

File: db/config.cc [hunk 9]
```diff
@@ -849,6 +849,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , override_decommission(this, "override_decommission", value_status::Used, false, "Set true to force a decommissioned node to join the cluster (cannot be set if consistent-cluster-management is enabled")
     , enable_repair_based_node_ops(this, "enable_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, true, "Set true to use enable repair based node operations instead of streaming based")
     , allowed_repair_based_node_ops(this, "allowed_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, "replace,removenode,rebuild,bootstrap,decommission", "A comma separated list of node operations which are allowed to enable repair based node operations. The operations can be bootstrap, replace, removenode, decommission and rebuild")
+    , enable_compacting_data_for_streaming_and_repair(this, "enable_compacting_data_for_streaming_and_repair", liveness::LiveUpdate, value_status::Used, true, "Enable compacting the data for streaming and repair (load'n'stream included). Can reduce the amount of data to be processed, but adds CPU overhead.")
```
Comment (asias) [inline db/config.cc:852]: We can make it clear this is option is not about the normal compaction or off-startegy compaction after repair or streaming.

How about:

Enable the compacting reader which compacts data before sending or synchronizing  data to the peer node. This allows skipping sending data covered by tombstones and reducing the amount of data to be processed, but adds CPU overhead.

**Review by asias — APPROVED** [review]
Looks good. I have a minor comment about the description of the new option.

File: db/config.cc (hunk 9 above)
Comment (denesb) [reply db/config.cc]: Done in v4.

**Comment by denesb** [comment]
v4:
* Further clarify config item's description.

File: test/boost/mutation_reader_test.cc (hunk 7 above)
Comment (denesb) [reply test/boost/mutation_reader_test.cc]: See https://github.com/scylladb/scylla-dtest/pull/3321

File: replica/database.hh (hunk 1 above)
Comment (avikivity) [reply replica/database.hh:692]: An expired tombstone could just as well have been dropped a minute before streaming started due to compaction. It's not reliable. If there are problems with resurrection, they should be solved; streaming expired tombstones doesn't solve anything.

**Comment by denesb** [comment]
v5:
* Fix the compacting-reader's next-partition() support. It caused a validation error if there was an open range tombstone change.
* Log error which caused the repair to fail.

Topology tests seem to be flaky with this series.  I need to look into it.

**Review by asias — APPROVED** [review]
Looks good. Thanks Botond!

File: repair/repair.cc [hunk 10]
```diff
@@ -589,9 +589,13 @@ repair::shard_repair_task_impl::shard_repair_task_impl(tasks::task_manager::modu
 void repair::shard_repair_task_impl::check_failed_ranges() {
     rlogger.info("repair[{}]: stats: repair_reason={}, keyspace={}, tables={}, ranges_nr={}, {}",
         global_repair_id.uuid(), _reason, _status.keyspace, table_names(), ranges.size(), _stats.get_stats());
-    if (nr_failed_ranges || _aborted || _failed) {
-        auto msg = format("repair[{}]: {} out of {} ranges failed, keyspace={}, tables={}, repair_reason={}, nodes_down_during_repair={}, aborted_by_user={}",
-                global_repair_id.uuid(), nr_failed_ranges, ranges_size(), _status.keyspace, table_names(), _reason, nodes_down, _aborted);
+    if (nr_failed_ranges || _aborted || _failed_because) {
+        sstring failed_because = "N/A";
+        if (!_aborted) {
+            failed_because = _failed_because ? *_failed_because : "uknown";
```
[the remark this answers is not in the document]
Comment (bhalevy) [reply repair/repair.cc:595]: I kinda like "uknown", sounds nice :)

**Comment by denesb** [comment]
v6:
* Fixed typo s/uknown/unknown/
* Drop patch corutinizing `insert_repair_meta()`, it makes `topology_custom` tests flaky.

File: repair/repair.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (denesb) [reply repair/repair.cc]: Done in v5.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2713/

Out of space. Re-kicked.

**Comment by denesb** [comment]
Related: https://github.com/scylladb/scylladb/issues/9882
This issue has only one thing left in its TODO list: migrating all usages of Java `sstable*` to scylla native alternatives. Once done, we can close it.

**Comment by tgrabiec** [comment]
> * Drop patch corutinizing `insert_repair_meta()`, it makes `topology_custom` tests flaky.

Do you know why?

**Comment by denesb** [comment]
> @tgrabiec - #14860 perhaps

Yes, I that is the issue I opened for it.

**Comment by denesb** [comment]
> > @tgrabiec - #14860 perhaps
> 
> Yes, I that is the issue I opened for it.

In any case, this is orthogonal to this PR. We should investigate it separately.
